### DISCUSSION TITLE
How to implement a ParamConverterProvider that works with same key query process

### DISCUSSION THREAD

**Discussion by Sboti03** [eclipse-ee4j/jersey#5686] [Q&A]
Hi everyone,

I'm facing an issue with handling collection type `@QueryParam` in my JAX-RS application. I'm using `ParamConverterProvider` to convert query parameters to custom Java types. However, I'm having trouble when the request includes multiple keys for the same parameter.

When I call the endpoint with multiple keys like this:

```
/api?name=Bob&name=James
```

Only the first parameter (`Bob`) gets converted. However, it works fine when I use a single key with comma-separated values:

```
/api?name=Bob,James
```

Here's a brief overview of my setup:
- I'm using `ParamConverterProvider` to provide `ParamConverter` instances for custom types.
- I've created a `CollectionParameterConverter` to handle collections of these types.
- The conversion works as expected for comma-separated values but not for multiple keys.

How can I properly handle a collection type `@QueryParam` when the request includes multiple keys for the same parameter? Any guidance or examples on how to adapt the `ParamConverterProvider` or other relevant parts of the code would be greatly appreciated.

Thanks for your help!

---

**Comment by MILLERMARRU** [comment]
I traced this through Jersey's `MultivaluedParameterExtractorFactory` and the root cause is that registering a `ParamConverter` for your collection type directly opts you out of Jersey's built-in multi-key aggregation entirely.

```java
ParamConverter<?> converter = paramConverterFactory.getConverter(rawType, type, annotations);
if (converter != null) {
    return new SingleValueExtractor(converter, parameterName, defaultValue);
}

// Check whether the rawType is the type of the collection supported.
if (rawType == List.class || rawType == Set.class || rawType == SortedSet.class) {
    // ...
    return CollectionExtractor.getInstance(rawType, converter, parameterName, defaultValue);
}
```

Jersey checks for a converter matching your parameter's raw type first, and that branch wins if it matches. The comment right above it in the source is explicit about why: "if someone writes a converter that support List\<Integer\> this approach should precede the next one." Since your `ParamConverterProvider` returns a `ParamConverter<YourCollectionType>` for the exact type of your `@QueryParam` field, Jersey wires up a `SingleValueExtractor`, which by design extracts a single string value out of the multivalued query map (the first `name=` it sees) and hands that one string to your converter. The `List`/`Set`/`SortedSet` branch with `CollectionExtractor`, the one that actually aggregates every repeated `name=` occurrence before conversion, never gets reached because the first branch already returned.

The fix is to stop putting the converter on the collection type and put it on the element type instead:

```java
@GET
public Response search(@QueryParam("name") List<Name> names) { ... }
```

```java
public class NameParamConverterProvider implements ParamConverterProvider {
    @Override
    public <T> ParamConverter<T> getConverter(Class<T> rawType, Type genericType, Annotation[] annotations) {
        if (rawType != Name.class) return null; // element type, not the List
        return (ParamConverter<T>) new NameParamConverter();
    }
}
```

With the converter registered for `Name` (the element type) rather than `List<Name>`, Jersey's raw-type check on the field falls through to the `List.class` branch, resolves the generic argument, finds your converter for `Name`, and uses `CollectionExtractor`, which correctly collects every `name=` occurrence from the query string and calls your converter once per value. That gets you `?name=Bob&name=James` working the way you expect. If you still want comma-splitting supported too, you can layer that inside the converter's own `fromString` for the single-value case, but the multi-key case specifically needs the extractor path, not a converter sitting on the wrapper collection type.
